File: server/modules/agents/match.ts
import { AGENT_ADAPTERS } from "./adapters.js";
import type { AgentAdapter, AgentKind } from "./types.js";

export function extractInvocation(commandLine: string): string | null {
  let trimmed = commandLine.trim();
  if (!trimmed) return null;

  for (let i = 0; i < 8; i++) {
    const withoutAssignments = trimmed.replace(/^(?:\s*[A-Za-z_][A-Za-z0-9_]*=\S*\s*)+/, "");
    if (withoutAssignments !== trimmed) {
      trimmed = withoutAssignments.trimStart();
      continue;
    }
    if (trimmed.startsWith("env ")) {
      trimmed = trimmed.slice(4).trimStart();
      continue;
    }
    break;
  }

  const match = trimmed.match(/^([^\s|;&]+)/);
  if (!match) return null;
  const token = match[1]!;
  return token.includes("/") ? token.slice(token.lastIndexOf("/") + 1) : token;
}

export function matchAgentAdapter(commandLine: string): AgentAdapter | null {
  const invocation = extractInvocation(commandLine);
  if (!invocation) return null;
  return AGENT_ADAPTERS.find((adapter) => adapter.binaries.includes(invocation)) ?? null;
}

export function getAgentAdapter(kind: AgentKind): AgentAdapter | undefined {
  return AGENT_ADAPTERS.find((adapter) => adapter.kind === kind);
}

export function buildAgentResumeArgv(
  kind: AgentKind,
  sessionId: string,
): string[] | null {
  const adapter = getAgentAdapter(kind);
  if (!adapter) return null;
  return adapter.resumeArgs(sessionId);
}

File: server-rs/src/agents/match.rs
use regex::Regex;
use std::sync::OnceLock;

use super::adapters::agent_adapters;
use super::types::{AgentAdapter, AgentKind};

fn env_assignments_re() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| {
        Regex::new(r"^(?:\s*[A-Za-z_][A-Za-z0-9_]*=\S*\s*)+").expect("env assignments regex")
    })
}

pub fn extract_invocation(command_line: &str) -> Option<String> {
    let mut current = command_line.trim().to_string();
    if current.is_empty() {
        return None;
    }

    for _ in 0..8 {
        let without_assignments = env_assignments_re().replace(&current, "");
        if without_assignments.as_ref() != current.as_str() {
            current = without_assignments.trim_start().to_string();
            continue;
        }
        if let Some(rest) = current.strip_prefix("env ") {
            current = rest.trim_start().to_string();
            continue;
        }
        break;
    }

    let token = current.split([' ', '|', ';', '&']).next()?.trim();
    if token.is_empty() {
        return None;
    }
    if let Some(idx) = token.rfind('/') {
        Some(token[idx + 1..].to_string())
    } else {
        Some(token.to_string())
    }
}

pub fn match_agent_adapter(command_line: &str) -> Option<&'static dyn AgentAdapter> {
    let invocation = extract_invocation(command_line)?;
    for adapter in agent_adapters() {
        if adapter.binaries().contains(&invocation.as_str()) {
            return Some(adapter);
        }
    }
    None
}

pub fn get_agent_adapter(kind: AgentKind) -> Option<&'static dyn AgentAdapter> {
    for adapter in agent_adapters() {
        if adapter.kind() == kind {
            return Some(adapter);
        }
    }
    None
}

pub fn build_agent_resume_argv(kind: AgentKind, session_id: &str) -> Option<Vec<String>> {
    get_agent_adapter(kind).map(|adapter| adapter.resume_args(session_id))
}
